lib/concurrent/promise.rb

```diff
@@ -243,6 +240,26 @@ def execute
       self
     end
 
+    def set(value = IVar::NO_VALUE, &block)
```

Comment:
@pitr-ch This version of `set` follows your suggestion and calls `execute`. The function itself could use a little cleanup but the tests still pass. I think the behavior here is probably more accurate. I haven't changed `fail` yet but I anticipate it will be very similar except for something like this:

```ruby
@promise_body = Proc.new { |result| raise reason }
```

Thoughts?